### DISCUSSION TITLE
Why does generic type variance affect nullable analysis?

### DISCUSSION THREAD

**Discussion by lorcanmooney** [dotnet/roslyn#60519] [Q&A] [answered]
I wouldn't have expected the `out`-ness of a type parameter to affect nullable analysis, yet it seems to introduce a spurious warning in the code below. Invariant and `in` parameters don't trigger the warning. Am I missing something?

```csharp
#nullable enable

using System;
using System.Diagnostics.CodeAnalysis;

class Foo<T> : IFoo<T> { }

interface IFoo<out T> { }    // Remove 'out' and the CS8603 warning below goes away

static class FooExt
{
    public static bool TryGetFoo<T>(this IFoo<T> a, [MaybeNullWhen(false)] out T b)
    {
        throw new NotImplementedException();
    }
}

class FooTest
{
    string Test()
    {
        IFoo<string> foo = new Foo<string>();

        if (foo.TryGetFoo(out string? f1))
        {
            return f1;    // CS8603: Possible null reference return.
        }

        throw new NotImplementedException();
    }
}
```

**Comment by RikkiGibson** [comment, marked as answer]
Type parameter variance matters for nullable analysis as far as it determines when conversions are valid and whether certain signature differences across overrides/implementations are valid. For example, for `IFoo<out T>`, we can convert an `IFoo<string>` to `IFoo<string?>`. With any other variance such a conversion would be a warning.

However, it seems like the variance shouldn't affect whether a warning is given at `return f1` here. I think some investigation in method type argument inference would be needed. My "hypothesis" would be that for some reason we are inferring a nullable type argument for `T` in `TryGetFoo<T>` only when the variance of `T` in `IFoo<T>` is `out`.

We might already have some workaround for the usage of `[MaybeNullWhen]` here, which somehow doesn't account for the variance--basically, it feels like the type of an argument for a parameter with such an annotation shouldn't necessarily contribute its nullability to the type inference.

**Reply by RikkiGibson** [reply]
I opened #60726 since I do think there is a bug here.
